h2o-py/h2o/estimators/estimator_base.py

```diff
@@ -228,6 +228,7 @@ def _make_parms(self, x=None, y=None, training_frame=None, offset_column=None, f
             raise H2OValueError("Verbose mode is not available for %s" % self.__class__.__name__)
         parms = self._parms.copy()
         names = training_frame.names if training_frame is not None else []
+        names_set = set(names)
```

Comment:
This is gonna work but there is no guarantee it won't happen again, accidental use of names would create a regression.

Would it be possible to change the representation of names to OrderedDict to emulate the ordered set?

Reply:
true, even if this is not a part of client code that is supposed to change often.

if you just want to be able to access values by index, `OrderedDict` doesn't add much value compared with a simple dict, e.g.
```python
names = dict(zip(range(len(training_frame.names)), training_frame.names)
...
name = names[i]
...
if n in names.values()
...
```
but then we can't use `in names`, just `in names.values()`, for the first syntax we would need our own data structure...
Probably not urgent though, I can create a follow-up ticket

Reply:
also dict is very slow...
1st version with dict, 2nd with pure set, only using 10k cols as 100k takes too long

```python
In [37]: def f():
    ...:     cols = [f"C{n}" for n in range(10000)]
    ...:     start_p = time.time()
    ...:     cols = dict(zip(range(len(cols)), cols))
    ...:     x = set(cols.values()) - {"C1234"}
    ...:     end_p = time.time() - start_p
    ...:     start_l = time.time()
    ...:     for c in x:
    ...:         if c in cols.values():
    ...:             pass
    ...:     end_l = time.time() - start_l
    ...:     return end_p, end_l
    ...:

In [38]: f()
Out[38]: (0.0015070438385009766, 1.0008137226104736)

In [39]: def f():
    ...:     cols = [f"C{n}" for n in range(10000)]
    ...:     start_p = time.time()
    ...:     cols = set(cols)
    ...:     x = cols - {"C12345"}
    ...:     end_p = time.time() - start_p
    ...:     start_l = time.time()
    ...:     for c in x:
    ...:         if c in cols:
    ...:             pass
    ...:     end_l = time.time() - start_l
    ...:     return end_p, end_l
    ...:

In [40]: f()
Out[40]: (0.0009548664093017578, 0.0006127357482910156)
```

Reply:
I can try this, it's pretty fast:

```python
In [53]: class LookupList(list):
    ...:     def __init__(self, l):
    ...:         super().__init__(l)
    ...:         self._set = set(l)
    ...:     def __contains__(self, item):
    ...:         return item in self._set
    ...:     def __sub__(self, other):
    ...:         return self._set - other
    ...:     def __add__(self, other):
    ...:         return self._set + other
    ...:

In [58]: def f():
    ...:     cols = [f"C{n}" for n in range(10000)]
    ...:     start_p = time.time()
    ...:     cols = LookupList(cols)
    ...:     x = cols - {"C1234"}
    ...:     end_p = time.time() - start_p
    ...:     start_l = time.time()
    ...:     for c in x:
    ...:         if c in cols:
    ...:             pass
    ...:     end_l = time.time() - start_l
    ...:     return end_p, end_l
    ...:

In [59]: f()
Out[59]: (0.0014090538024902344, 0.0023229122161865234)
```

Reply:
sounds like a good idea to encapsulate it



also on another note: Would there be a better built-in structure in Python 3.x+? Even though we support Python 2.7, we don't have to necessarily address this use case as well - it is an edge case. If we had a fast solution for 3.x I think that would be okay as well (as long as we document the Py 2.7 shortcomings)

Reply:
@michalkurka just pushed a read-only `LookupList` optimized for lookup as its name suggests.

I couldn't find any Py3 data structure that provides both random access and fast "in" lookup, so I guess it makes sense to have our own DS. Mine would still be slow if I was looking for element index for example... so it's a very specific kind of optimization, probably why this is not part of the standard library.